src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.cs

```diff
@@ -203,22 +203,60 @@ string GetElementTypeFQN() => valueType is { } vt
     private static string FormatComparerOptions(ImmutableEquatableArray<CollectionConstructorParameter> parameters)
         => $"global::PolyType.Abstractions.CollectionComparerOptions.{FormatOptionsComparerPropertyName(parameters) ?? "None"}";
 
-    private string FormatAssociatedTypeShapes(TypeShapeModel objectShapeModel)
+    private static string? GetAssociatedTypesFactoryName(TypeShapeModel typeShapeModel)
     {
-        if (objectShapeModel.AssociatedTypes.Count == 0)
-        {
-            return "null";
-        }
+        return typeShapeModel.AssociatedTypes.Count > 0 ? $"__GetAssociatedTypes_{typeShapeModel.SourceIdentifier}" : null;
+    }
 
-        StringBuilder builder = new();
-        builder.Append("static associatedType => associatedType switch { ");
+    private void FormatAssociatedTypesFactory(SourceWriter writer, TypeShapeModel objectShapeModel, string factoryName)
+    {
+        Debug.Assert(objectShapeModel.AssociatedTypes.Count > 0);
+
+        writer.WriteLine($$"""
+            private global::PolyType.ITypeShape? {{factoryName}}(global::System.Type associatedType)
```

Comment:
@AArnott heads up, I've also updated the associated shapes get method to use the same generation strategy as ITypeShapeProvider.